shared/yaml/codeql/serverless/ServerLess.qll

```diff
@@ -0,0 +1,164 @@
+   */
+  bindingset[base]
+  private string removeTrailingDot(string base) {
+    if base.regexpMatch(".*/\\.")
+    then result = base.substring(0, base.length() - 1)
+    else result = base
+  }
+
+  /**
+   * Gets a string where a leading "./" is simplified to "" (if it exists).
+   */
+  bindingset[base]
+  private string removeLeadingDotSlash(string base) {
+    if base.regexpMatch("\\./.*") then result = base.substring(2, base.length()) else result = base
+  }
+
+  /**
+   * Gets a string suitable as part of a file path.
+   *
+   * Maps the empty string to the empty string.
+   */
+  bindingset[base]
+  private string normalizePath(string base) {
+    result = removeLeadingDotSlash(removeTrailingDot(base))
+  }
+
+  /**
+   * Holds if the `.yml` file `ymlFile` contains a serverless configuration from `framework` with
+   * `handler`, `codeURI`, and `runtime` properties.
+   * `codeURI` and `runtime` default to the empty string if no explicit value is set in the configuration.
+   *
+   * `handler` should be interpreted in a language specific way, see `mapping.md`.
+   */
+  predicate hasServerlessHandler(
+    File ymlFile, string framework, string handler, string codeUri, string runtime
+  ) {
+    exists(YamlMapping resource | ymlFile = resource.getFile() |
+      // There exists at least "AWS::Serverless::Function" and "Aliyun::Serverless::Function"
+      resource.lookup("Type").(YamlScalar).getValue().regexpMatch(".*::Serverless::Function") and
+      framework = lookupValue(resource, "Type") and
+      exists(YamlMapping properties | properties = resource.lookup("Properties") |
+        (
+          handler = lookupValue(properties, "Handler") and
+          codeUri = normalizePath(lookupValueOrEmpty(properties, "CodeUri")) and
+          runtime = lookupValueOrEmpty(properties, "Runtime")
+        )
+      )
+      or
+      // The `serverless` library, which specifies a top-level `functions` property
+      framework = "Serverless" and
+      exists(YamlMapping functions |
+        functions = resource.lookup("functions") and
+        not exists(resource.getParentNode()) and
+        handler = lookupValue(functions.getValue(_), "handler") and
+        codeUri = "" and
+        runtime = lookupValueOrEmpty(functions, "Runtime")
+      )
+    )
+  }
+
+  /**
+   * Holds if `handler` = `filePart . astPart` and `filePart` does not contain a `.`.
+   * This is a convenience predicate, as in many cases the first part of the handler property
+   * should be interpreted as (the stem of) a file name.
+   */
+  bindingset[handler]
+  predicate splitHandler(string handler, string filePart, string astPart) {
+    exists(string pattern | pattern = "(.*?)\\.(.*)" |
+      filePart = handler.regexpCapture(pattern, 1) and
+      astPart = handler.regexpCapture(pattern, 2)
+    )
+  }
+
+  /**
+   * Holds if a file with stem `fileStem` has a serverless handler denoted by `func`.
```

Comment:
We use the term "stem" to refer to the basename without extension (see `getStem` in `Files.qll`), but here it refers to the absolute path without extension.

Could you rename variables to avoid this inconsistency, perhaps by renaming to something like `path` or `pathNoExt`?